# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [36]:
# Import your libraries:
import sklearn as s

# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [29]:
# Your code here:
diabetes=s.datasets.load_diabetes()

Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [30]:
# Your code here:
diabetes.keys()

dict_keys(['data', 'target', 'frame', 'DESCR', 'feature_names', 'data_filename', 'target_filename'])

#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [16]:
# Your code here:
print(diabetes["DESCR"])

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - age     age in years
      - sex
      - bmi     body mass index
      - bp      average blood pressure
      - s1      tc, T-Cells (a type of white blood cells)
      - s2      ldl, low-density lipoproteins
      - s3      hdl, high-density lipoproteins
      - s4      tch, thyroid stimulating hormone
      - s5      ltg, lamotrigine
      - s6      glu, blood sugar level

Note: Each of these 10 feature va

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?
    - Each of these 10 feature variables
    
1. What is the relation between `diabetes['data']` and `diabetes['target']`?
    
    - predictor (data) and response (target) variables. 
    
1. How many records are there in the data?
    - Number of Instances: 442

#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [27]:
# Your code here:
diabetes["data"].shape,diabetes["target"].shape

((442, 10), (442,))

# Challenge 2 - Perform Supervised Learning on the Dataset

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### Let's briefly revisit the linear regression formula:

```
y = β0 + β1X1 + β2X2 + ... + βnXn + ϵ
```

...where:

- X1-Xn: data 
- β0: intercept 
- β1-βn: coefficients 
- ϵ: error (cannot explained by model)
- y: target

Also take a look at the `sklearn.linear_model.LinearRegression` [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html).

#### In the cell below, import the `linear_model` class from `sklearn`. 

In [46]:
import numpy as np
from sklearn.linear_model import LinearRegression

#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [58]:
# Your code here:
diabetes.keys()

diabetes_model=LinearRegression()
diabetes_model

LinearRegression()

#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

In [54]:
# Your code here:
diabetes_data_train=diabetes["data"][:-20]
diabetes_target_train=diabetes["target"][:-20]
diabetes_data_test=diabetes["data"][-20:]
diabetes_target_test=diabetes["target"][-20:]


Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [65]:
# Your code here:
reg=diabetes_model.fit(diabetes_data_train,diabetes_target_train)
print(reg.intercept_)
print(reg.coef_)

152.76430691633442
[ 3.03499549e-01 -2.37639315e+02  5.10530605e+02  3.27736980e+02
 -8.14131709e+02  4.92814588e+02  1.02848452e+02  1.84606489e+02
  7.43519617e+02  7.60951722e+01]


#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [73]:
# Your code here:
predict=reg.predict(diabetes_data_test)
predict

array([197.61846908, 155.43979328, 172.88665147, 111.53537279,
       164.80054784, 131.06954875, 259.12237761, 100.47935157,
       117.0601052 , 124.30503555, 218.36632793,  61.19831284,
       132.25046751, 120.3332925 ,  52.54458691, 194.03798088,
       102.57139702, 123.56604987, 211.0346317 ,  52.60335674])

#### Print your `diabetes_target_test` and compare with the prediction. 

In [98]:
# Your code here:
import pandas as pd
df = pd.DataFrame({"predict":predict})
df["gt"]=diabetes_target_test
df

predict     gt
0   197.618469  233.0
1   155.439793   91.0
2   172.886651  111.0
3   111.535373  152.0
4   164.800548  120.0
5   131.069549   67.0
6   259.122378  310.0
7   100.479352   94.0
8   117.060105  183.0
9   124.305036   66.0
10  218.366328  173.0
11   61.198313   72.0
12  132.250468   49.0
13  120.333292   64.0
14   52.544587   48.0
15  194.037981  178.0
16  102.571397  104.0
17  123.566050  132.0
18  211.034632  220.0
19   52.603357   57.0

#### Is `diabetes_target_test` exactly the same as the model prediction? Explain.

# Your explanation here:
No, hay error entre ambos.

# Bonus Challenge 1 - Hypothesis Testing with `statsmodels`

After generating the linear regression model from the dataset, you probably wonder: then what? What is the statistical way to know if my model is reliable or not?

Good question. We'll discuss that using Scikit-Learn in Challenge 5. But for now, let's use a fool-proof way by using the ([Linear Regression class of StatsModels](https://www.statsmodels.org/dev/regression.html)) which can also conduct linear regression analysis plus much more such as calcuating the F-score of the linear model as well as the standard errors and t-scores for each coefficient. The F-score and t-scores will tell you whether you can trust your linear model.

To understand the statistical meaning of conducting hypothesis testing (e.g. F-test, t-test) for slopes, read [this webpage](https://onlinecourses.science.psu.edu/stat501/node/297/) at your leisure time. We'll give you a brief overview next.

* The F-test of your linear model is to verify whether at least one of your coefficients is significantly different from zero. Translating that into the *null hypothesis* and *alternative hypothesis*, that is:

    ```
    H0 : β1 = β2 = ... = β10 = 0
    HA : At least one βj ≠ 0 (for j = 1, 2, ..., 10)
    ```

* The t-tests on each coefficient is to check whether the confidence interval for the variable contains zero. If the confidence interval contains zero, it means the null hypothesis for that variable is not rejected. In other words, this particular vaiable is not contributing to your linear model and you can remove it from your formula.

Read the documentations of [StatsModels Linear Regression](https://www.statsmodels.org/dev/regression.html) as well as its [`OLS` class](https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html) which stands for *ordinary least squares*.

#### In the next cell, analyze `diabetes_data_train` and `diabetes_target_train` with the linear regression model of `statsmodels`. Print the fit summary.

Your output should look like:

![statsmodels regression](../statsmodels.png)

In [113]:
# Your code here:
import numpy as np
import statsmodels.api as sm
mod = sm.OLS()
res = mod.fit()
res.summary()

ValueError: shapes (422,10) and (422,10) not aligned: 10 (dim 1) != 422 (dim 0)

In [118]:
pd.DataFrame(diabetes_data_train,diabetes_target_train)

0         1         2         3         4         5         6  \
151.0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
75.0  -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
141.0  0.085299  0.050680  0.044451 -0.005671 -0.045599 -0.034194 -0.032356   
206.0 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
135.0  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   
...         ...       ...       ...       ...       ...       ...       ...   
55.0  -0.052738 -0.044642  0.071397 -0.074528 -0.015328 -0.001314  0.004460   
84.0   0.009016 -0.044642 -0.024529 -0.026328  0.098876  0.094196  0.070730   
42.0  -0.020045 -0.044642 -0.054707 -0.053871 -0.066239 -0.057367  0.011824   
146.0  0.023546 -0.044642 -0.036385  0.000068  0.001183  0.034698 -0.043401   
212.0  0.038076  0.050680  0.016428  0.021872  0.039710  0.045032 -0.043401   

              7         8         9  
151.0 -0.002592  0.019908 -0.017646  
75.0  -0.039493 -0.068330 -0.092204  
141.0 -0.002592  0.002864 -0.025930  
206.0  0.034309  0.022692 -0.009362  
135.0 -0.002592 -0.031991 -0.046641  
...         ...       ...       ...  
55.0  -0.021412 -0.046879  0.003064  
84.0  -0.002592 -0.021394  0.007207  
42.0  -0.039493 -0.074089 -0.005220  
146.0  0.034309 -0.033249  0.061054  
212.0  0.071210  0.049769  0.015491  

[422 rows x 10 columns]

#### Interpreting hypothesis testing results

Answer the following questions in the cell below:

1. What is the F-score of your linear model and is the null hypothesis rejected?

1. Does any of the t-tests of the coefficients produce a confidence interval containing zero? What are they?

1. How will you modify your linear reguression model according to the test results above?

In [ ]:
# Your answers here:

# Challenge 3 - Peform Supervised Learning on a Pandas Dataframe

Now that we have dealt with data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [147]:
# Your code here:
import pandas as pd
auto=pd.read_csv("../auto-mpg.csv")


Look at the first 5 rows using the `head()` function:

In [148]:
# Your code here:
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [128]:
# Your code here:
auto.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horse_power   392 non-null    float64
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model_year    398 non-null    int64  
 7   car_name      398 non-null    object 
dtypes: float64(4), int64(3), object(1)
memory usage: 25.0+ KB


What is the newest model year and the oldest model year?

In [132]:
# Your code here:

auto.model_year.max(),auto.model_year.min()

(82, 70)

Check the dataset for missing values and remove all rows containing at least one missing value.

In [161]:
# Your code here:
auto.dropna(how='any',inplace=True)

In [162]:
auto.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392 entries, 0 to 397
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horse_power   392 non-null    float64
 4   weight        392 non-null    int64  
 5   acceleration  392 non-null    float64
 6   model_year    392 non-null    int64  
dtypes: float64(4), int64(3)
memory usage: 24.5 KB


Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [139]:
# Your code here:

auto.cylinders.value_counts()

4    199
8    103
6     83
3      4
5      3
Name: cylinders, dtype: int64

In [149]:
auto.drop(columns=["car_name"],inplace=True)

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

*Hint: To separate data for training and test, use the `train_test_split` method we used in previous labs.*

In [159]:
auto

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0    18.0          8         307.0        130.0    3504          12.0   
1    15.0          8         350.0        165.0    3693          11.5   
2    18.0          8         318.0        150.0    3436          11.0   
3    16.0          8         304.0        150.0    3433          12.0   
4    17.0          8         302.0        140.0    3449          10.5   
..    ...        ...           ...          ...     ...           ...   
393  27.0          4         140.0         86.0    2790          15.6   
394  44.0          4          97.0         52.0    2130          24.6   
395  32.0          4         135.0         84.0    2295          11.6   
396  28.0          4         120.0         79.0    2625          18.6   
397  31.0          4         119.0         82.0    2720          19.4   

     model_year  
0            70  
1            70  
2            70  
3            70  
4            70  
..          ...  
393          82  
394          82  
395          82  
396          82  
397          82  

[398 rows x 7 columns]

In [164]:
x=auto.drop(columns=["mpg"])
y=auto.mpg

In [165]:
from sklearn.model_selection import train_test_split

In [166]:
# Your code here:
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2)

Now we will processed and peform linear regression on this data to predict the mpg for each vehicle. 

#### In the next cell, create an instance of the linear regression model and call it `auto_model`. Fit `auto_model` with your training data.

In [172]:
# Your code here:
auto_model=s.linear_model.LinearRegression()
auto_model.fit(X_train,y_train)


LinearRegression()

# Challenge 4 - Evaluate the Model

In addition to evaluating your model with F-test and t-test, you can also use the *Coefficient of Determination* (a.k.a. *r squared score*). This method does not simply tell *yes* or *no* about the model fit but instead indicates how much variation can be explained by the model. Based on the r squared score, you can decide whether to improve your model in order to obtain a better fit.

You can learn about the r squared score [here](). Its formula is:

![R Squared](../r-squared.png)

...where:

* yi is an actual data point.
* ŷi is the corresponding data point on the estimated regression line.

By adding the squares of the difference between all yi-ŷi pairs, we have a measure called SSE (*error sum of squares*) which is an application of the r squared score to indicate the extent to which the estimated regression model is different from the actual data. And we attribute that difference to the random error that is unavoidable in the real world. Obviously, we want the SSE value to be as small as possible.

#### In the next cell, compute the predicted *y* based on `X_train` and call it `y_pred`. Then calcualte the r squared score between `y_pred` and `y_train` which indicates how well the estimated regression model fits the training data.

*Hint: r squared score can be calculated using `sklearn.metrics.r2_score` ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)).*

In [177]:
# Your code here:
from sklearn.metrics import r2_score
y_pred=auto_model.predict(X_train)
r2_score(y_train,y_pred)

0.8193490017528257

#### Our next step is to evaluate the model using the test data. 

We would like to ensure that our model is not overfitting the data. This means that our model was made to fit too closely to the training data by being overly complex. If a model is overfitted, it is not generalizable to data outside the training data. In that case, we need to reduce the complexity of the model by removing certain features (variables).

In the cell below, use the model to generate the predicted values for the test data and assign them to `y_test_pred`. Compute the r squared score of the predicted `y_test_pred` and the oberserved `y_test` data.

In [178]:
# Your code here:
y_test_pred=auto_model.predict(X_test)
r2_score(y_test,y_test_pred)


0.7564405858988966

#### Explaining the results

The r squared scores of the training data and the test data are pretty close (0.8146 vs 0.7818). This means our model is not overfitted. However, there is still room to improve the model fit. Move on to the next challenge.

# Challenge 5 - Improve the Model Fit

While the most common way to improve the fit of a model is by using [regularization](https://datanice.github.io/machine-learning-101-what-is-regularization-interactive.html), there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [181]:
# Your code here:

X_train09, X_test09,y_train09,y_test09= train_test_split(x,y, train_size=0.9)

Initialize a new linear regression model. Name this model `auto_model09`. Fit the model to the new sample (training) data.

In [182]:
# Your code here:
auto_model09=LinearRegression()
auto_model09.fit(X_train09,y_train09)

LinearRegression()

Compute the predicted values and r squared score for our new model and new sample data.

In [183]:
# Your code here:

# Your code here:
auto_model09.score(X_train09,y_train09)

0.8076224698811318

Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [185]:
# Your code here:
auto_model09.score(X_test09,y_test09)

0.8017127289008749

# Bonus Challenge 2 - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE

In [ ]:
from sklearn.feature_selection import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`

In [ ]:
# Your code here:


Fit the model and print the ranking

In [ ]:
# Your code here:


Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [ ]:
# Your code here:


Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [ ]:
# Your code here: 


#### Conclusion

You may obtain the impression from this lab that without knowing statistical methods in depth, it is difficult to make major progress in machine learning. That is correct. If you are motivated to become a data scientist, statistics is the subject you must be proficient in and there is no shortcut. 

Completing these labs is not likely to make you a data scientist. But you will have a good sense about what are there in machine learning and what are good for you. In your future career, you can choose one of the three tracks:

* Data scientists who need to be proficient in statistical methods.

* Data engineers who need to be good at programming.

* Data integration specialists who are business or content experts but also understand data and programming. This cross-disciplinary track brings together data, technology, and business and will be in high demands in the next decade.